core/src/main/java/com/datastrato/gravitino/catalog/CatalogEventDispatcher.java

```diff
@@ -0,0 +1,149 @@
+import com.datastrato.gravitino.exceptions.NoSuchMetalakeException;
+import com.datastrato.gravitino.listener.EventBus;
+import com.datastrato.gravitino.listener.api.event.AlterCatalogEvent;
+import com.datastrato.gravitino.listener.api.event.AlterCatalogFailureEvent;
+import com.datastrato.gravitino.listener.api.event.CreateCatalogEvent;
+import com.datastrato.gravitino.listener.api.event.CreateCatalogFailureEvent;
+import com.datastrato.gravitino.listener.api.event.DropCatalogEvent;
+import com.datastrato.gravitino.listener.api.event.DropCatalogFailureEvent;
+import com.datastrato.gravitino.listener.api.event.ListCatalogEvent;
+import com.datastrato.gravitino.listener.api.event.ListCatalogFailureEvent;
+import com.datastrato.gravitino.listener.api.event.ListCatalogInfoEvent;
+import com.datastrato.gravitino.listener.api.event.ListCatalogInfoFailureEvent;
+import com.datastrato.gravitino.listener.api.event.LoadCatalogEvent;
+import com.datastrato.gravitino.listener.api.event.LoadCatalogFailureEvent;
+import com.datastrato.gravitino.listener.api.info.CatalogInfo;
+import com.datastrato.gravitino.utils.PrincipalUtils;
+import java.util.Map;
+
+/**
+ * {@code CatalogEventDispatcher} is a decorator for {@link CatalogDispatcher} that not only
+ * delegates catalog operations to the underlying catalog dispatcher but also dispatches
+ * corresponding events to an {@link EventBus} after each operation is completed. This allows for
+ * event-driven workflows or monitoring of catalog operations.
+ */
+public class CatalogEventDispatcher implements CatalogDispatcher {
+  private final EventBus eventBus;
+  private final CatalogDispatcher dispatcher;
+
+  /**
+   * Constructs a CatalogEventDispatcher with a specified EventBus and CatalogDispatcher.
+   *
+   * @param eventBus The EventBus to which events will be dispatched.
+   * @param dispatcher The underlying {@link CatalogDispatcher} that will perform the actual catalog
+   *     operations.
+   */
+  public CatalogEventDispatcher(EventBus eventBus, CatalogDispatcher dispatcher) {
+    this.eventBus = eventBus;
+    this.dispatcher = dispatcher;
+  }
+
+  @Override
+  public NameIdentifier[] listCatalogs(Namespace namespace) throws NoSuchMetalakeException {
+    try {
+      NameIdentifier[] nameIdentifiers = dispatcher.listCatalogs(namespace);
+      eventBus.dispatchEvent(new ListCatalogEvent(PrincipalUtils.getCurrentUserName(), namespace));
+      return nameIdentifiers;
+    } catch (Exception e) {
+      eventBus.dispatchEvent(
+          new ListCatalogFailureEvent(PrincipalUtils.getCurrentUserName(), e, namespace));
+      throw e;
+    }
+  }
+
+  @Override
+  public Catalog[] listCatalogsInfo(Namespace namespace) throws NoSuchMetalakeException {
+    try {
+      Catalog[] catalogs = dispatcher.listCatalogsInfo(namespace);
+      eventBus.dispatchEvent(
+          new ListCatalogInfoEvent(PrincipalUtils.getCurrentUserName(), namespace));
```

Comment:
seems no need to create a sperate ` ListCatalogInfoEvent`, we could reuse `ListCatalogEvent`, @jerryshao @mchades WDYT?

Reply:
I'm OK with this solution.

Reply:
removed `ListCatalogInfoEvent`